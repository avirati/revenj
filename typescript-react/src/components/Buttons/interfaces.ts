export interface ISimpleButton {
  className?: string;
  disabled?: boolean;
  label: string;
  url?: string;
  isExternalUrl?: boolean;
  tooltip?: string;
  onClick?: () => void;
}

export interface IComponentButton {
  Component: React.ComponentType<{}>;
}

export type ButtonItem = ISimpleButton | IComponentButton;

export const isComponentButton = (it: any): it is IComponentButton =>
  it != null && (it as IComponentButton).Component != null;
