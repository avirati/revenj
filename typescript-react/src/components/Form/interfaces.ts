export enum FormType {
  View = 'View',
  Create = 'Create',
  Edit = 'Edit',
}

export interface IFormError {
  message: string | JSX.Element;
  code?: string | number;
}

export enum ErrorType {
  Error,
  Warning,
  QAPink,
}
