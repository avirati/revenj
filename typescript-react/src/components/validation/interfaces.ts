import { Numeric } from '../../util/NumberUtils/NumberUtils';

// Operator argument types
export type Requireable = any;
export type Ord = Numeric;
export type Num = Numeric;
export type Eq = any;

// Operator interfaces
export interface IOperator<B, I> {
  (baseValue: B): (inputValue: I) => boolean;
}
export interface IRequireableOperator extends IOperator<undefined, Requireable> {
  (): (inputValue: Requireable) => boolean;
}

export interface IOrdOperator extends IOperator<Ord, Ord> {
  (baseValue: Numeric): (inputValue: Numeric) => boolean;
}

export interface IOrdAlphaOperator extends IOperator<string, string> {}

export interface INumOperator extends IOperator<Ord, Ord> {
  (baseValue: Numeric): (inputValue: Numeric) => boolean;
}

export interface IEqOperator extends IOperator<any, any> {
  <T>(baseValue: T): (inputValue: T) => boolean;
}

export interface IGetValidatorBaseValue<B, F, P> {
  (allValues: F, componentProps: P, fieldName: string): B;
}

export interface IGetValidatorErrorMessage<B, I, F, P> {
  (validationBaseValue: B, inputValue: I, allValues: F, componentProps: P, fieldName: string): string;
}

export interface IParseValue<R, V> {
  (input: R): V;
}

export type ValidationError = string | undefined;
export type Validator<I, F, P> = (inputValue: I, allValues?: F, props?: P, name?: keyof F) => ValidationError;

export enum ValidateTrigger {
  Always = 'Always',
  IfInputIsSet = 'IfInputIsSet',
  IfBaseIsSet = 'IfBaseIsSet',
  IfInputAndBaseAreSet = 'IfInputAndBaseAreSet',
}

interface IValidatorCreatorProps<B, I, F, P> {
  validateOverride?: ValidateTrigger;
  getValidatorBaseValueOverride?: IGetValidatorBaseValue<B, F, P> | B;
  getValidatorErrorMessageOverride?: IGetValidatorErrorMessage<B, I, F, P> | string;
  getValueFromInputOverride?: IParseValue<any, I>;
  getValueFromBaseOverride?: IParseValue<any, B>;
}

export type ValidatorCreator<B, I, F, P> = (props?: IValidatorCreatorProps<B, I, F, P>) => Validator<I, F, P>;

export type Operator<B, I> = (baseValue: B) => (inputValue: I) => boolean;

export interface IFormFieldSpecification {
  label: string;
  path: string;
}

// In many cases, the path is not actually required
export interface IFormFieldSpecificationDescriptor extends Omit<IFormFieldSpecification, 'path'> {
  path?: IFormFieldSpecification['path'];
}

export interface IFormFieldValue<T> extends IFormFieldSpecification {
  value: T;
}

export interface ITransformer<V, R> {
  (value: V): R;
}

export type RegexOperator = (regex: RegExp) => (inputValue: string) => boolean;

export type Preprocessor<T> = (value: T) => T;
