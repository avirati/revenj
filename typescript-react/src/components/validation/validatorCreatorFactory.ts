import { identity } from '../../util/FunctionalUtils/FunctionalUtils';
import {
  IGetValidatorBaseValue,
  IGetValidatorErrorMessage,
  IParseValue,
  Operator,
  Preprocessor,
  ValidateTrigger,
  ValidatorCreator,
} from './interfaces';

const globalPreProcessors: Array<Preprocessor<any>> = [];

export const withPreprocessor = <T>(preprocessor: Preprocessor<T>) =>
  globalPreProcessors.push(preprocessor);

// we treat undefined, null and '' as an absence of values
export const valueIsAbsent = (value: any) =>
  (value == null) || (typeof value === 'string' && value === '');

////////////////////////////////////////////////////////////////////////////////
// Validator creator                                                          //
////////////////////////////////////////////////////////////////////////////////
// Generic types: B = base type, I = input type, F = form type, P = props type
export interface IValidatorCreatorOptions<B, I, F, P> {
  validate?: ValidateTrigger;
  getValidatorErrorMessage: IGetValidatorErrorMessage<B, I, F, P> | string;
  getValidatorBaseValue?: IGetValidatorBaseValue<B, F, P> | B;
  getValueFromInput?: IParseValue<any, I>;
  getValueFromBase?: IParseValue<any, B>;
  operator: Operator<B, I>;
}
export const validatorCreatorFactory = <B, I, F, P>({
  validate = ValidateTrigger.IfInputIsSet,
  getValidatorErrorMessage,
  getValidatorBaseValue,
  getValueFromInput = (identity as IParseValue<I, I>),
  getValueFromBase = (identity as IParseValue<B, B>),
  operator,
}: IValidatorCreatorOptions<B, I, F, P>): ValidatorCreator<B, I, F, P> =>
  (creatorProps = {}) =>
    (inputValue, allValues, props, name) => {
      const {
        validateOverride,
        getValidatorBaseValueOverride,
        getValidatorErrorMessageOverride,
        getValueFromInputOverride,
        getValueFromBaseOverride,
      } = creatorProps;

      const validateIf = validateOverride || validate;
      const getBaseValue = getValidatorBaseValueOverride || getValidatorBaseValue;
      const parseInputValue: IParseValue<any, I> = getValueFromInputOverride || getValueFromInput;
      const parseBaseValue: IParseValue<any, B> = getValueFromBaseOverride || getValueFromBase;

      const baseValue = typeof getBaseValue === 'function'
        ? (getBaseValue as IGetValidatorBaseValue<B, F, P>)(allValues ?? {} as F, props!, name as string)
        : getBaseValue;

      if (
        ((validateIf === ValidateTrigger.IfInputIsSet) && (valueIsAbsent(inputValue))) ||
        ((validateIf === ValidateTrigger.IfBaseIsSet) && (valueIsAbsent(baseValue))) ||
        ((validateIf === ValidateTrigger.IfInputAndBaseAreSet) && (valueIsAbsent(inputValue) || valueIsAbsent(baseValue)))
      ) {
        return undefined;
      }

      const parsedBaseValue = globalPreProcessors.reduce((val, f) => f(val), parseBaseValue(baseValue));
      const parsedInputValue = globalPreProcessors.reduce((val, f) => f(val), parseInputValue(inputValue));

      const getErrorMessage = getValidatorErrorMessageOverride || getValidatorErrorMessage;
      const message = typeof getErrorMessage === 'function'
        ? getErrorMessage(parsedBaseValue, parsedInputValue, allValues ?? {} as F, props!, name as string)
        : getErrorMessage;

      return operator(parsedBaseValue)(parsedInputValue)
        ? undefined
        : message;
    };
