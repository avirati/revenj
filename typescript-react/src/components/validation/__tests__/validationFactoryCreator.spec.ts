import { ValidateTrigger } from '../interfaces';
import { validatorCreatorFactory, valueIsAbsent } from '../validatorCreatorFactory';

describe('validation', () => {
  describe('valueIsAbsent', () => {
    it('should return true for undefined, null and empty string, true otherwise', () => {
      expect(valueIsAbsent(undefined)).toBe(true);
      expect(valueIsAbsent(null)).toBe(true);
      expect(valueIsAbsent('')).toBe(true);

      expect(valueIsAbsent(0)).toBe(false);
      expect(valueIsAbsent(1)).toBe(false);
      expect(valueIsAbsent(false)).toBe(false);
      expect(valueIsAbsent(true)).toBe(false);
      expect(valueIsAbsent('a string')).toBe(false);
      expect(valueIsAbsent([])).toBe(false);
      expect(valueIsAbsent({})).toBe(false);
    });
  });

  describe('validatorCreatorFactory', () => {
    it('should create a valid validator creator with a validation error message', () => {
      interface IInterval {
        min: number;
        max: number;
      }

      const intervalidatorCreator = validatorCreatorFactory<IInterval, number, void, void>({
        getValidatorBaseValue: () => ({ min: 10, max: 20 }),
        getValidatorErrorMessage: () => `Input value is outside of the interval`,
        operator: (interval) => (input) =>
          input > interval.min && input < interval.max,
      });

      const dayInWeek = intervalidatorCreator({ getValidatorBaseValueOverride: { min: 1, max: 7 } });

      expect(dayInWeek(2)).toBeUndefined();
      expect(dayInWeek(0)).toBe('Input value is outside of the interval');
    });

    it('should create a valid validator creator with a validation error message function', () => {
      interface IInterval {
        min: number;
        max: number;
      }

      const intervalidatorCreator = validatorCreatorFactory<IInterval, number, void, void>({
        getValidatorBaseValue: () => ({ min: 10, max: 20 }),
        getValidatorErrorMessage: ({ min, max }, input) => `Input value {${input}} is not from interval <${min}, ${max}>`,
        operator: (interval) => (input) =>
          input > interval.min && input < interval.max,
      });

      const dayInWeek = intervalidatorCreator({ getValidatorBaseValueOverride: { min: 1, max: 7 } });

      expect(dayInWeek(2)).toBeUndefined();
      expect(dayInWeek(0)).toBe('Input value {0} is not from interval <1, 7>');
    });

    it('should not catch the error from getValidatorErrorMessage if it throws', () => {
      const failidatorCreator = validatorCreatorFactory<void, any, void, void>({
        getValidatorErrorMessage: () => { throw new Error('It indeed failed!'); },
        operator: () => () => true,
        validate: ValidateTrigger.Always,
      });

      const failidator = failidatorCreator();
      const fail = () => failidator('dummy');
      expect(fail).toThrowErrorMatchingSnapshot();
    });

    it('should not catch the error from getValueFromInput if it throws', () => {
      const failidatorCreator = validatorCreatorFactory<void, any, void, void>({
        getValidatorErrorMessage: 'It failed!',
        getValueFromInput: () => { throw new Error('It indeed failed!'); },
        operator: () => () => true,
        validate: ValidateTrigger.Always,
      });

      const failidator = failidatorCreator();
      const fail = () => failidator('dummy');
      expect(fail).toThrowErrorMatchingSnapshot();
    });

    it('should not catch the error from getValidatorBaseValue if it throws', () => {
      const failidatorCreator = validatorCreatorFactory<void, any, void, void>({
        getValidatorBaseValue: () => { throw new Error('It indeed faild because reasons!'); },
        getValidatorErrorMessage: 'It failed!',
        operator: () => () => true,
        validate: ValidateTrigger.Always,
      });

      const failidator = failidatorCreator();
      const fail = () => failidator('dummy');
      expect(fail).toThrowErrorMatchingSnapshot();
    });

    it('should not catch the error from getValueFromBase if it throws', () => {
      const failidatorCreator = validatorCreatorFactory<void, any, void, void>({
        getValidatorErrorMessage: 'It failed!',
        getValueFromBase: () => { throw new Error('It indeed failed!'); },
        operator: () => () => true,
        validate: ValidateTrigger.Always,
      });

      const failidator = failidatorCreator();
      const fail = () => failidator('dummy');
      expect(fail).toThrowErrorMatchingSnapshot();
    });
  });
});
