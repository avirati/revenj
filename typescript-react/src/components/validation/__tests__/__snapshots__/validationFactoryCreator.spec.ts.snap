// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation validatorCreatorFactory should not catch the error from getValidatorBaseValue if it throws 1`] = `"It indeed faild because reasons!"`;

exports[`validation validatorCreatorFactory should not catch the error from getValidatorErrorMessage if it throws 1`] = `"It indeed failed!"`;

exports[`validation validatorCreatorFactory should not catch the error from getValueFromBase if it throws 1`] = `"It indeed failed!"`;

exports[`validation validatorCreatorFactory should not catch the error from getValueFromInput if it throws 1`] = `"It indeed failed!"`;
